src/network/admin_server.rs

```diff
@@ -178,4 +231,22 @@ impl AdminService for MyAdminService {
             Err(err) => Err(Status::from_error(Box::new(err))),
         }
     }
+
+    async fn upload_snapshot(
+        &self,
+        _request: Request<Empty>,
+    ) -> std::result::Result<Response<Empty>, Status> {
+        let now = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .map_err(|err| Status::from_error(Box::new(err)))?
+            .as_millis();
+
+        self.backup_shard(0, self.block_store.db.clone(), now as i64)
+            .await?;
+        for (shard, stores) in self.shard_stores.iter() {
+            self.backup_shard(*shard, stores.db.clone(), now as i64)
+                .await?
+        }
```

Comment:
Is this happening in the context of a request? Might want to return immediately and do the actual backup in a background thread.

Also, we should add protection for multiple requests. It shouldn't cause multiple backups (maybe just check for the presence of the backup dir?).

Reply:
I added protection for parallel requests. 

I'm not sure about returning early-- then we can't return the error back to the client. Maybe we want to turn this into a streaming rpc? I'll hold off on making this change-- if we really want it we can do it later.